imap/jmap_notif.c

```diff
@@ -128,12 +149,31 @@ static int append_eventnotif(const char *from,
     struct buf buf = BUF_INITIALIZER;
     const char *type = json_string_value(json_object_get(jnotif, "type"));
     const char *ical_uid = json_string_value(json_object_get(jnotif, "calendarEventId"));
+    int notifications_max = config_getint(IMAPOPT_JMAP_MAX_CALENDAREVENTNOTIFS);
+    if (notifications_max < 0) notifications_max = 0;
 
+    // Prune notifications each time we create a new one.
+    if (notifications_max) {
+        prune_notifications(notifmbox, (size_t) notifications_max);
+    }
+
+    // Expunge all former notifications for destroyed events.
     if (!strcmp(type, "destroyed")) {
-        /* Expunge all former event notifications for this UID */
-        struct mailbox_iter *iter = mailbox_iter_init(notifmbox, 0, 0);
+        struct mailbox_iter *iter = mailbox_iter_init(notifmbox, 0,
+                ITER_STEP_BACKWARD|ITER_SKIP_UNLINKED|
+                ITER_SKIP_EXPUNGED|ITER_SKIP_DELETED);
+        size_t n_notifications = 0;
+
         message_t *msg;
         while ((msg = (message_t *) mailbox_iter_step(iter))) {
+
+            // Notifications exceeding notifications_max must
+            // have been pruned already, no need to check.
```

Comment:
This seems pointless - since they've been pruned, they're going to be skipped by ITER_SKIP_EXPUNGED - so we have changed the direction of iteration for no reason

Reply:
(I guess it does stop reading potentially thousands if something wrote a ton of notifications in the cyr_expire window

Reply:
I also had a good hard think about this part while I was reviewing it, and eventually concluded that it was sensible even though on the surface it seemed redundant.  I've forgotten my reasoning though... probably should have written it down! 😅